Review {cfr} v0.1.0 for CRAN release

pratikunterwegs opened the pull request:
This PR is for a full package review prior to a CRAN release, and is also working towards items in the release checklist in #68.

This PR will be used to collect and address reviwer comments. Either this branch, or issue-specific branches will be used to push changes. The changes will be merged into `main` prior to release.

The proposed timeline is for all reviews to be in by **8th September 2023**. The goal is to have {cfr} released to CRAN soon after. This schedule is contingent on releasing {epiparameter} to CRAN.

I would encourage/request two reviewers to sign up to do the full review. Please tick a box and assign yourself in the Reviewers panel on the right to indicate that you have taken on the review.

- [x] Reviewer 1
- [x] Reviewer 2

Following review this PR also fixes #80 and fixes #81.

pratikunterwegs commented:
> Thanks to @abdoelnaser-mahmood-degoot for sending me these comments over Slack - posting them here for all others to read.

I went through the {cfr} package and have some comments that I'd want to share with you.

1- I see an entanglement between two arguments: `correct_for_delays` and `epidist`, across several functions in the package resulting from how you designed/implemented the code. If `correct_for_delays` is set to `TRUE`, then `epidist` must be specified. The question is, what are the chances that the user will provide a value for the `epidist` parameter while ignoring to correct for the delays? When a user enters a value for `epidist`, he or she implicitly intends to account for delays. The `correct_for_delays` parameter may aid in the design and implementation of the package, but it may confuse the user.
2- In the quickest example, remove the comma after the data argument on the `rolling_cfr_naive` function.
3- In the `readme`, it would be better to offer the reader a brief code chunk that generates the `ggplot2` figure.
4- Include a link to the source `Camacho et al.` in the documentation of `ebola1976` data.
5- Mention the data type of `severity_baseline` and `severity_baseline` in the documentation of the `estimate_ascertainment` function parameters, just as you did for the other parameters. Also, what is the default value for the `type` argument, and why does `max_date` accept a string instead of a `Date`?
6- The default value for poisson_threshold, in many functions, is `100` and not `200`.
7- Is it correct that you set the default value for the `epidist` argument to NULL in the `known_outcomes` and `estimate_ascertainment` functions? Likewise, for `severity_baseline` to `0.014`; is this true for all diseases?

Bisaloo commented on R/estimate_ascertainment.R:
```diff
@@ -0,0 +1,147 @@
+  type <- match.arg(type, several.ok = FALSE)
+
+  # create empty data frame
+  df_severity <- data.frame()
```
I see this pattern in a couple of places in the code. What's the rationale behind it? It doesn't seem to serve any purpose and is not idiomatic in R.

pratikunterwegs replied:
This and the one other instance has now been removed.

Bisaloo replied:
There is one last occurrence here: https://github.com/epiverse-trace/cfr/blob/90d787f5a36f6e18da08250694827f509df3434b/R/estimate_severity.R#L40

Bisaloo commented on R/estimate_ascertainment.R:
```diff
@@ -0,0 +1,147 @@
+  )
+
+  # return data with columns in correct order
+  df_out[, c("ascertainment_me", "ascertainment_lo", "ascertainment_hi")]
```
As far as I can tell, the reordering will have no effect since the column names where created to be precisely in this order.

pratikunterwegs replied:
No, the column names are created `c("ascertainment_me", "ascertainment_hi", "ascertainment_lo")`; the reordering flips the order of low and high to make it consistent with other outputs.

Bisaloo commented on DESCRIPTION:
```diff
@@ -0,0 +1,84 @@
+Imports:
+    checkmate,
+    stats
+Suggests: 
+    dplyr,
+    tidyr,
+    purrr,
+    scales,
+    forcats,
+    ggplot2,
+    pak,
+    testthat (>= 3.0.0),
+    usethis,
+    knitr,
+    bookdown,
+    rmarkdown,
+    spelling,
+    epiparameter,
+    incidence2,
+    outbreaks
```
From a quick glance, it doesn't look like all those packages need to be listed here. In particular, pak and usethis probably shouldn't be here. Could you review this?

pratikunterwegs replied:
Removed pak, usethis, and rmarkdown, which I think are the only extra packages here. @jamesmbaazam has suggested replacing Tidyverse packages with {tidyverse}, but I would prefer to have them listed separately.

pratikunterwegs commented:
Thanks @Bisaloo, will wait for other reivews to come in and move to fix in the week of 11th Sept.

adamkucharski commented on vignettes/estimate_ascertainment.Rmd:
````diff
@@ -0,0 +1,294 @@
+```
+
+```{r }
+df_reporting_varying <- estimate_ascertainment(
````
Looks like the problem is related to the importation of epiparameter object:
```
> onset_to_death_covid <- epidist_db(
+   disease = "COVID-19",
+   epi_dist = "onset_to_death",
+   author = "Linton_etal"
+ )
Returning 2 results that match the criteria (2 are parameterised). 
Use subset to filter by entry variables or single_epidist to return a single entry. 
To retrieve the short citation for each use the 'get_citation' function
```

pratikunterwegs replied:
Thanks, this is the same as the issue above, flagged in #80.

adamkucharski commented on vignettes/estimate_time_varying_severity.Rmd:
````diff
@@ -0,0 +1,372 @@
+We retrieve the appropriate distribution reported in @linton2020 using the `epidist_db()` function from the _epiparameter_ package.
+
+```{r message=FALSE, warning=FALSE, eval=TRUE}
+onset_to_death_covid <- epidist_db(
````
I get same issue as in `estimate_ascertainment.Rmd` with epidist being a list. I think I'm using latest version of epiparameter, but may be missed a refactoring somewhere?

pratikunterwegs replied:
Yep, same as above.

adamkucharski commented on vignettes/estimate_time_varying_severity.Rmd:
```diff
@@ -0,0 +1,372 @@
+\end{equation}
+
+We then assume that the severity measure, for example CFR, of interest is
+Binomially-distributed, in the following way 
```
```suggestion
binomially distributed, in the following way 
```

adamkucharski commented on vignettes/estimate_time_varying_severity.Rmd:
```diff
@@ -0,0 +1,372 @@
+  d_t \sim \text{Binomial}(k_t, \theta_t). 
+\end{equation}
+
+We use maximum-likelihood techniques to determine the value of $\theta_t$ for 
+each $t$, whereby $\theta$ represents the severity measure of interest.
```
```suggestion
We use maximum likelihood estimation to determine the value of $\theta_t$ for 
each $t$, where $\theta$ represents the severity measure of interest.
```

adamkucharski reviewed (commented):
Looking good overall. There are few issues I had with pulling from epiparameter – it may be a version problem, but wanted to flag.

General comment: the documentation uses a mix of 'case fatality rate', 'case fatality risk' and 'case fatality ratio', sometimes within the same document (e.g. vignette). We should pick one for consistency. 'Case fatality rate' is most commonly used, even though it is technically inaccurate (the value calculated is a risk rather than a rate). I would therefore go with 'case fatality risk' (or 'case fatality ratio' if this requires less refactoring).

Similarly, 'ascertainment rate' isn't really a rate - perhaps 'ascertainment ratio' or 'ascertainment probability' would be better?

jamesmbaazam reviewed (commented):
# {cfr} full package review

I think {cfr} is in good shape and will be ready to be submitted to CRAN following this full package review. I have added some comments below. Please let me know if you have any questions.

## Routine automatic checks

The following routine checks were performed on the package and did not raise any issues.

- [ ] `devtools::check()`
- [ ] `goodpractice::gp()`
- [ ] `lintr::lint_package()`
- [ ] `styler::style_pkg`
- [ ] `covr::package_coverage()`

## README

### General comments

- The README looks fine to me. It's clear and concise, and all the badges are appropriate.

### Improvements

- I have added line specific comments to the README file.
- I suggest to provide a brief overview of the core functions and what they do. This will help users to quickly understand what the package does and how to use it.

## Functions

### General comments

- `estimate_static()`:
  - One major comment is for you to add details on the methods and references for the CFR estimation. This is important for people who are not familiar with CFR and want to learn more about it. Same comment goes for `known_outcomes()`.
  - Additionally, I would suggest adding a `_cfr` suffix to this function and others, i.e. `estimate_static_cfr()`. This will help users to quickly identify the functions that estimate CFRs.
-  `known_outcomes()`:
- `estimate_ascertainment()`: I would be interested to see the methods for this function as I am a little worried about its inner workings. For example, its results are highly sensitive to the `severity_baseline` argument, which is user supplied rather than estimated. I also have a little quibble about replacing all ascertainment ratios > 1 with 1. It seems hacky and I wonder why it is not dealt with by the model but rather as a post-processing step.

### Improvements

- I have added line specific comments to each function.

## Vignettes

### General comments

- The vignettes lay out the core functionality in the package and are insightful. I have added line specific comments.
- As a general comment, since vignettes are basically literate programming, I usually prefer not to use redundant phrases like "using the code below" or "the code below", "the following code", etc. The accompanying code should reinforce the written words.

#### Specific comments

- "Estimating the proportion of cases that are ascertained during an outbreak":
  - Another way to visualize the results to remove the effect of the `severity_baseline` argument is to plot the relative ratios between the countries so that the baseline severity cancels out, ruling out its effect.
- "Calculating a static, delay-adjusted estimate of disease severity":
  - I'm not convinced that Fig 3 and 7 are necessary. It's not clear what it adds to the text. I would suggest to remove it. It looks like something that belongs in {epiparameter} rather than {cfr}.
  - I would suggest moving the section "Details: Adjusting for delays between two time series" to a theory vignette like in [bpmodels](https://epiverse-trace.github.io/bpmodels/articles/theoretical_background.html) and the details section of the respective functions. To be honest, it looks good here and seems to fit the context. Another way to go about it would be to hyperlink to the theory vignette at the beginning and end of this vignette like you have for the Getting Started vignette.
  - I would suggest replacing the generic use of "known outcomes" with the specifics. I think in this vignette, it is the known death outcomes.
- "Estimating how disease severity varies over the course of an outbreak":
  - Title seems long. Suggest shortening to, for example, "Estimating time-varying CFRs"
  - Same as before, I'm not convinced that Fig 3 is necessary. It's not clear what it adds to the text. I would suggest to remove it.
  - Suggest adding a legend to Fig 4 and 5 indicate what the grey and tomato lines/ribbons represent.
  - Also, what is going on before April in Figs 4 and 5? The ribbons look weird and are hard to interpret.
  - Refer to my earlier comment above on moving the section "Details: Adjusting for delays between two time series".

### Improvements

- I have added line specific comments to each vignette.

## Other infrastructure

### DESCRIPTION

Package version is missing the patch number, i.e., currently `0.1` instead of `0.1.0`. I would suggest to addding it to conform with the rest of our packages.

### Tests

- Might be worthwhile to provide more descriptive contexts for the tests, for example, `test_that("multiplication works", ...`.
- I did not inspect the tests in detail so do not have detailed comments.

joshwlambert commented:
## Full package review for {cfr}.

Overall the package looks good. It is functionally to the point which should help new users get to grips with the aims and applications of the package. The documentation is clearly written and provides a comprehensive overview of the functions and data requirements. 

I like that the package is minimal in terms of namespace but contains a good amount of functionality. The inclusion of two datasets provides two clear use-cases for the package.

I've left some comments below. Some of these overlap with those already made by other reviewers but I thought I'd leave them in just to ensure they weren't missed.

### Functions
- Why does the `severity_baseline` argument in `estimate_ascertainment()` have a default of 0.014? This seems to me to be an argument you want to ensure the user specifies to prevent accidentally leaving the default value.
- In `estimate_ascertainment()` is the `apply()` on L127 needed. From what I can see it is will always be a single row data.frame and can also have vectorised operations if it has more than one row. 
- In `estimate_ascertainment()` (and other functions) there seems to be an interaction of function arguments which is not documentation or caught by input checking. The setup appears to give `type` precedence so if `type = "static"` and an `<epidist>` is supplied it is assumed the user wants to use the `"static"` estimate. Not sure if this is a problem, and if it is I'm not sure I have a solution. One option (not sure how I feel about as it removes some of the clarity/verbosity of the function arguments) is to remove the `type` argument and instead control the flow of the function with `is.null(epidist)`.
- I think he first part of `stopifnot()` in `estimate_rolling()` (L61) is already checked by `checkmate::assert_names()` on L54-57
- In `estimate_static()` is there a reason `severity_estimate` is assigned as an empty numeric? It seems that the `if else` that follows means that `severity_estimate` will always be assigned before the function returns
- In `estimate_time_varying()` there are two input checks on `correct_for_delays`
- The ordering of arguments seems to varying across the `estimate_*()` functions. Would be good to provide users with consistent argument ordering
- `estimate_time_varying()` has a `# TODO input checking` comment but it seems that all input is checked
- `epidist` is checked twice in input checking in `estimate_time_varying()`
- In `get_default_burn_in()` it is possible to have an `<epidist>` object with a mean of `NA`. I would try and account for this. If you think it would be useful I can write an S3 method for `mean()` (thankfully it's a S3 generic in base) which can try and calculate a mean even if it is not given in the summary stats of the object.
- The `epidist` default in `known_outcomes` is `NULL`. However, if an `epidist` is not supplied the function will error on L60 at `assert_class(epidist, "epidist")`. I would suggest removing the argument default.
- The `prepare_data.incidence2()` function definition contains `...`. I suggest adding `chkDots(...)` to the function which will prevent users supplying arguments to be silently swallowed by `...`. I would also document that the dots are not used as currently the documentation is somewhat misleading and suggests these arguments will be used by the method.


### Documentation
- `estimate_ascertainment()` and `estimate_time_varying()` example fails due to changes in {epiparameter}. Adding `single_epidist = TRUE` to `epidist_db()` will solve the issue
- It would be good to add the full reference of Camacho et al. (2014) to the `ebola1976.R` file, as you did for the `covid_data.R` file.
- Check consistent use of imperative function description and full sentences (e.g. `estimate_ascertainment.R`)
- Inconsistent use of `data frame`, `data.frame` & `<data.frame>` in documentation
- Is `estimate_rolling()` using a method described or used in a paper, if so could it be referenced in the documentation?
- `known_outcomes` can `@inheritParams` from `estimate_static()` for `@param epidist`
- My assumption is that when users want to read the documentation for a method they use `?generic`, in this case `?prepare_data`, as this is the function they are interacting with and the dispatch takes place under the hood. I think the documentation for `prepare_data.incidence2()` is better and should be merged with the documentation for the generic so the user can find it all in one place. As this generic currently only has one method I would reword the documentation as it currently states: "it has methods for classes commonly found in epidemiological data", instead you can say that more methods for common epidemiological data classes will be added over time. 
- Due to documenting both the generic and the method for `prepare_data()` it appears twice in the package reference on the pkgdown site. If you want to keep both that is perfectly fine. I think the usual practise is to not document the method and just have the generic in the reference (e.g [dplyr](https://dplyr.tidyverse.org/reference/filter.html)), but if you plan to add more methods you may want each of these to be shown. It can be quiet useful to have the class name in the function to show which classes the function applies to, in this case document the methods and just export the generic.
- The code for the plot in the README is completely hidden. Is it worth having it hidden but with an option to reveal the code if users are interested in using it?
- Is there a reason for the double asterisk in the figure caption of Fig 3 in `estimate_static_severity.Rmd`
- Why does `library(outbreaks)` require a `# nolint` flag in the `data_from_incidence2.Rmd` vignette?

### Testing
- Testing coverage for all functions is high except for `prepare_data()` which has 72.41% coverage.

### Package infrastructure
- update GitHub action workflows to match those in {packagetemplate}
- Add citation (.cff) file, CITATION file, and update-citation-cff.yml workflow, (this will require an addition to the Rbuildignore which is automatic if I remember correctly)
- There are some folders in the .gitignore that are missing from the .gitignore from the {packagetemplate} and vice versa, would be good to check if these should be added to.
- Update `.lintr` to match {packagetemplate}
- In data-raw you save the ebola data with `usethis::use_data(..., version = 3)` and save the covid data with `usethis::use_data(..., version = 2)`. Is there a reason for the difference?
- A styler run updates a few files

_Speculative idea_: what if `prepare_data()` was moved internally to the `estimate_*()` functions? The `estimate_*()` functions remain normal (i.e. not S3 generics) but can accept multiple data types in `data` and `prepate_data()` does the work for the user. The way I see it the benefits to this approach are: 1) one less exported function (`prepare_data()` becomes internal); 2) less documentation around wrangling different data sources as the user does not need to know all the details. Downsides to this approach is potentially more arguments in the `estimate_*()` functions. But the current defaults of `prepare_data()` would match the requires documented to have a `"cases"` and a `"deaths"` column name.

pratikunterwegs commented:
Thanks all for your feedback - I shall work on reconciling your suggestions and making changes this week. I would say the aim is to be ready for a CRAN submission as soon as possible, ideally within this week.

pratikunterwegs commented on vignettes/data_from_incidence2.Rmd:
```diff
@@ -0,0 +1,154 @@
+
+# `<incidence2>` objects from linelists
+
+We first load some linelist data for the 2014 Ebola outbreak provided with the _outbreaks_ package, and create an _incidence2_ object and view it.
```
The `ebola1976` data are aggregated, the {outbreaks} data provide a linelist which we want for this example.

pratikunterwegs commented on vignettes/estimate_ascertainment.Rmd:
````diff
@@ -0,0 +1,294 @@
+The other arguments are the same as those found in `estimate_time_varying()`. 
+
+```{r }
+df_reporting_static <- estimate_ascertainment(
````
Thanks for flagging this - this is related to recent changes to {epiparameter}. Issue #80 has flagged these for edits. Essentially, `epidist_db()` now returns a list rather than an automatically chosen `<epidist>`. I will also replace the parameter access with `epiparameter::get_parameter()`.

pratikunterwegs commented on R/estimate_ascertainment.R:
```diff
@@ -0,0 +1,147 @@
+#'   author = "Linton_etal"
+#' )
+#'
+#' estimate_ascertainment(
+#'   data = df_covid_uk,
+#'   epidist = onset_to_death_covid,
+#'   type = "varying",
+#'   severity_baseline = 0.014,
+#'   smooth_inputs = TRUE,
+#'   burn_in_value = 7L,
+#'   correct_for_delays = TRUE,
+#'   max_date = "2020-06-30"
+#' )
+#'
```
Thanks, yes this is the error pointed out above due to recent changes to `epiparameter::epidist_db()`, flagged for fixing in #80.

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+#' )
+#'
+#' # estimate time varying severity without correcting for delays
+#' cfr_time_varying <- estimate_time_varying(
+#'   data = df_covid_uk_subset,
+#'   smooth_inputs = TRUE,
+#'   burn_in_value = 7L,
+#'   correct_for_delays = FALSE
+#' )
+#' # View
+#' tail(cfr_time_varying)
+#'
+#' # estimate time varying severity while correcting for delays
+#' cfr_time_varying <- estimate_time_varying(
+#'   data = df_covid_uk_subset,
+#'   epidist = onset_to_death_covid,
+#'   smooth_inputs = TRUE,
+#'   burn_in_value = 7L,
+#'   correct_for_delays = TRUE
+#' )
+#' tail(cfr_time_varying)
+#'
```
Yes, this is the same issue as in #80

pratikunterwegs commented on README.Rmd:
```diff
@@ -0,0 +1,185 @@
+# Load the Ebola 1976 data provided with the package
+data("ebola1976")
+
+# read delay distribution for ebolavirus onset to death from {epiparameter}
```
I don't see one missing, but feel free to suggest a addition.

pratikunterwegs commented on R/estimate_static.R:
```diff
@@ -0,0 +1,141 @@
+#' approximation. Set to 200 by default.
+#'
+#' @return A data.frame with the MLE and 95% confidence interval of the
+#' severity estimates, named "severity_me", "severity_lo", and "severity_hi".
```
I had renamed these measures to full names in an earlier stage of its development, but that got overwritten. I have kept it as is for now. Would you prefer to see full names? E.g. "severity_low", "severity_median"?

pratikunterwegs replied:
I have accepted this suggestion and changed the names; more details below.

pratikunterwegs commented on R/estimate_static.R:
```diff
@@ -0,0 +1,141 @@
+
+    # calculating the maximum likelihood estimate and 95% confidence interval
+    # using the binomial likelihood function from Nishiura
+    severity_estimate <- estimate_severity(
+      total_cases = sum(df_corrected$cases, na.rm = TRUE),
+      total_deaths = sum(df_corrected$deaths, na.rm = TRUE),
+      total_outcomes = sum(df_corrected$known_outcomes, na.rm = TRUE),
+      poisson_threshold = poisson_threshold
```
Thanks - could you take a look at [the details in the corresponding vignette](https://epiverse-trace.github.io/cfr/articles/estimate_static_severity.html#details-adjusting-for-delays-between-two-time-series) and the equivalent sections in the other vignettes, and check if this is what should be added to the function's _Details_?

pratikunterwegs commented on R/known_outcomes.R:
```diff
@@ -0,0 +1,101 @@
+#' case detection and death, typically approximated by a symptom onset to death
+#' distribution from the literature for the disease in question.
+#'
+#' @param data A data.frame containing the outbreak data. A daily time series
+#' with dates or some other absolute indicator of time (e.g. epiday/epiweek) and
+#' the numbers of new cases at each time point. This function does not require
+#' data on daily deaths, but this column (and any others) will be retained if
+#' present.
+#'
+#' @param epidist The delay distribution used, in the form of an
+#' [epiparameter::epidist()] object. This is used to obtain a probability
+#' mass function parameterised by time; i.e. \eqn{f(t)} which gives the
+#' probability a case has a known outcomes (i.e. death) at time \eqn{t},
+#' parameterised with disease-specific parameters before it is supplied here.
+#' A typical example would be a symptom onset to death delay distribution.
```
Thanks - I guess you mean `estimate_static()`, but I shall shift to using `@inheritParams`

pratikunterwegs commented on R/known_outcomes.R:
```diff
@@ -0,0 +1,101 @@
+#'  - "known_outcomes" for the total number of known outcomes on that day
+#' of the outbreak, and
+#'
+#'  - "u_t" for the under-reporting factor.
```
I think the details of the return data.frame would make more sense here, so we'll keep it for now.

pratikunterwegs commented on R/known_outcomes.R:
```diff
@@ -0,0 +1,101 @@
+#' )
+#'
+#' # examine the first few rows of the output
+#' head(
+#'   known_outcomes(data = ebola1976, onset_to_death_ebola)
+#' )
```
Thanks, this has been changed.

pratikunterwegs commented on R/known_outcomes.R:
```diff
@@ -0,0 +1,101 @@
+  )
+  checkmate::assert_class(epidist, "epidist")
+
+  pmf_vals <- stats::density(
+    epidist,
+    at = seq(from = 0, to = nrow(data) - 1L)
+  )
```
It is not an exported function from {epiparameter}, but {epiparameter} does implement it as a method for `<epidist>` objects, so I think this is the correct way to call this function.

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+#' Estimate known outcomes from case and death time-series data
```
Thanks for catching this, changed.

pratikunterwegs commented on R/estimate_static.R:
```diff
@@ -0,0 +1,141 @@
+  # check that data$date is a date column
+  checkmate::assert_date(data$date, any.missing = FALSE, all.missing = FALSE)
+  # check for excessive missing date and throw an error
+  stopifnot(
+    "Case data must contain columns `cases` and `deaths`" =
+      (all(c("cases", "deaths") %in% colnames(data))),
```
No, it was the same check - thanks for catching it, it has been removed.

pratikunterwegs commented on R/estimate_static.R:
```diff
@@ -0,0 +1,141 @@
+estimate_static <- function(data,
+                            correct_for_delays = FALSE,
+                            epidist = NULL,
+                            poisson_threshold = 100) {
```
The documentation has been corrected to read 100.

pratikunterwegs commented on R/estimate_static.R:
```diff
@@ -0,0 +1,141 @@
+#' )
+#'
+estimate_static <- function(data,
+                            correct_for_delays = FALSE,
```
I have accepted this suggestion and will add more details below.

pratikunterwegs commented on README.Rmd:
```diff
@@ -0,0 +1,185 @@
+
+_{{ packagename }}_ implements methods outlined in @nishiura2009, and CFR estimates based on more methods are likely to be added.
+
+_{{ packagename }}_ uses the [_epiparameter_ package](https://epiverse-trace.github.io/epiparameter/) for delay-corrected CFR estimates, and both packages are developed at the [Centre for the Mathematical Modelling of Infectious Diseases](https://www.lshtm.ac.uk/research/centres/centre-mathematical-modelling-infectious-diseases) at the London School of Hygiene and Tropical Medicine as part of the ...
```
This has been added now.

pratikunterwegs commented on R/estimate_ascertainment.R:
```diff
@@ -0,0 +1,147 @@
+#' @inheritParams estimate_time_varying
+#' @inheritParams estimate_static
+#'
+#' @param type A string, either `"static"` or `"varying"` which determines
```
I think the intention does come through so we're probably fine with allowing the shorter string for now.

pratikunterwegs commented on R/estimate_ascertainment.R:
```diff
@@ -0,0 +1,147 @@
+  }
+
+  # match argument for type
+  type <- match.arg(type, several.ok = FALSE)
```
Could you expand on this question, I haven't really understood it? Do you mean that the user might want both the static and the time-varying ascertainment? Would you expect those to be in the same data.frame?

jamesmbaazam replied:
I was thinking aloud about the case where a user wants both the static and time-varying results for some kind of comparison. I guess since they don't have the same dimensions, might make sense as a list of results.

pratikunterwegs replied:
I see - I think that's easy enough to do manually so let's leave it as it is. Happy to change is there's user demand

pratikunterwegs commented on R/estimate_ascertainment.R:
```diff
@@ -0,0 +1,147 @@
+    df_severity[, grepl("severity", colnames(df_severity), fixed = TRUE)],
+    MARGIN = 2,
+    FUN = function(x) {
+      x_ <- severity_baseline / x
+      x_[x_ > 1.0] <- 1.0
+      x_
```
What would you suggest as an alternative to this approach? This is likely to come up when data on deaths is poor, and the estimated severity is lower than the assumed baseline. I don't think returning values > 1.0 is an option, as I imagine users expect a ratio bounded between 0.0 and 1.0.

I could indeed throw a warning at this stage - would you suggest simply saying that ascertainment is > 1.0, or is it worth adding some epidemiological context as well? Or should that interpretation be left to the user?

pratikunterwegs commented:
Thanks to everyone who's chipped in. I've accepted in-line suggestions in two batches. I've addressed your suggestions one by one as well. I've also listed major changes below. Please do weigh in on the proposed changes too.
## User-facing changes

1. All `estimate_x()` functions now correct for delays by default; the previous behaviour was not to correct for delays. These functions now have the `epidist` argument set to missing by default, the previous default was `NULL`;
2. All `estimate_x()` functions now return values whose names take the full form, i.e., "severity_me" is now "severity_mean", "severity_lo" is "severity_low" etc.;
3. `estimate_static()` and `estimate_time_varying()` include method details from the vignettes.

@CarmenTamayo, we can coordinate re: the case study on Marburg in case you are using {cfr} already.
## Proposed user-facing changes
1. Rename CFR estimation functions to `cfr_x()` - `cfr_static()`, `cfr_rolling()`, `cfr_time_varying()`;
2. Make `known_outcomes()` an internal function as it does not appear to be intended for standalone use;
3. Remove the `correct_for_delays` argument, and make delay correction dependent on passing an `<epidist>`;
4. Relatedly, remove the `smooth_inputs` argument in `estimate_time_varying()` and make smoothing contingent on the value of `smoothing_window`.
## Internal changes
1. Simplifying function code, and removing duplicated checks;
2. Correcting how `get_standard_burn_in()` handles `<epidist>` objects;

## Scheduled changes
1. Updates to package DESCRIPTION following {epiparameter} CRAN release;
2. Updates to `get_default_burn_in()` to resolve issue #81.

pratikunterwegs commented:
## Suggestions and responses

@abdoelnaser-mahmood-degoot

>1- I see an entanglement between two arguments: `correct_for_delays` and `epidist`, across several functions in the package resulting from how you designed/implemented the code. If `correct_for_delays` is set to `TRUE`, then `epidist` must be specified. The question is, what are the chances that the user will provide a value for the `epidist` parameter while ignoring to correct for the delays? When a user enters a value for `epidist`, he or she implicitly intends to account for delays. The `correct_for_delays` parameter may aid in the design and implementation of the package, but it may confuse the user.

This is a fair point. @jamesmbaazam - you've suggested making correcting for delays the default, but this suggestion would conflict with that - do you have thoughts here as well?

>2- In the quickest example, remove the comma after the data argument on the `rolling_cfr_naive` function.

>3- In the `readme`, it would be better to offer the reader a brief code chunk that generates the `ggplot2` figure.

I think it's better to keep the Readme as short as possible and hide the plotting code. Similar examples are available in the vignettes.

>4- Include a link to the source `Camacho et al.` in the documentation of `ebola1976` data.

I have linked to the source using the `@source` tag, which is rendered as a link; see https://epiverse-trace.github.io/cfr/reference/ebola1976.html#source
>
>5- Mention the data type of `severity_baseline` and `severity_baseline` in the documentation of the `estimate_ascertainment` function parameters, just as you did for the other parameters. Also, what is the default value for the `type` argument, and why does `max_date` accept a string instead of a `Date`? 

I've added this to the documentation. The default value for `type` is `"static"`, as this is the first value in the vector passed to `type` in the function arguments. Ideally, we would expect users to explicitly pass the `type`. I have added a clarification to the documentation for this. The alternative is to not provide a vector of possible values as the default for `type`, and leave it missing. I think this makes it more difficult for users to quickly figure out the possible values, so I would propose retaining the vector of default values.

>6- The default value for poisson_threshold, in many functions, is `100` and not `200`.

Thanks - I have fixed the documentation here.

>7- Is it correct that you set the default value for the `epidist` argument to NULL in the `known_outcomes` and `estimate_ascertainment` functions? Likewise, for `severity_baseline` to `0.014`; is this true for all diseases?

Yes that's correct. Regarding `epidist`, this has been changed to provide no default value following @jamesmbaazam 's suggestion to make correcting for delays the default. See the discussion at the top.

pratikunterwegs commented:
@bisaloo

See comments on line suggestions.

pratikunterwegs commented:
@adamkucharski

> Looking good overall. There are few issues I had with pulling from epiparameter – it may be a version problem, but wanted to flag.

Thanks, yes, this is an issue (#80) due to changes in {epiparameter} that has been fixed now.

>General comment: the documentation uses a mix of 'case fatality rate', 'case fatality risk' and 'case fatality ratio', sometimes within the same document (e.g. vignette). We should pick one for consistency. 'Case fatality rate' is most commonly used, even though it is technically inaccurate (the value calculated is a risk rather than a rate). I would therefore go with 'case fatality risk' (or 'case fatality ratio' if this requires less refactoring).

Thanks - I've moved to using "case fatality risk" throughout.

>Similarly, 'ascertainment rate' isn't really a rate - perhaps 'ascertainment ratio' or 'ascertainment probability' would be better?

Thanks - I've moved to using "ascertainment ratio" throughout.

pratikunterwegs commented:
@jamesmbaazam

> - `estimate_static()`:
    - One major comment is for you to add details on the methods and references for the CFR estimation. This is important for people who are not familiar with CFR and want to learn more about it. Same comment goes for `known_outcomes()`.

Thanks, I have added the details for these methods from the corresponding vignettes. Could you take a look and say whether this is sufficient?

>- Additionally, I would suggest adding a `_cfr` suffix to this function and others, i.e. `estimate_static_cfr()`. This will help users to quickly identify the functions that estimate CFRs. 

Thanks for suggesting this - I'll bump this discussion to the top and ask for more feedback. I would prefer a naming convention of `cfr_x` where `x` is the method. This helps with communicating the functionality as well as with autocomplete.

>- `known_outcomes()`:

Could you add the suggestion here, you might have missed it.

>- `estimate_ascertainment()`: I would be interested to see the methods for this function as I am a little worried about its inner workings. For example, its results are highly sensitive to the `severity_baseline` argument, which is user supplied rather than estimated. I also have a little quibble about replacing all ascertainment ratios > 1 with 1. It seems hacky and I wonder why it is not dealt with by the model but rather as a post-processing step.

My understanding is that the function essentially calculates the appropriate severity measures, and then divides the 'baseline severity' by the measures. This is also why it's necessary to set values > 1 to be 1.0. Would values > 1.0 have some special interpretation, or simply be a mistake? In a suggestion elsewhere you said we should throw a warning for values > 1.0 - I have implemented this.

> As a general comment, since vignettes are basically literate programming, I usually prefer not to use redundant phrases like "using the code below" or "the code below", "the following code", etc. The accompanying code should reinforce the written words.

I agree with this and I thought I had removed these, but I've removed all remaining instances as well now - could you please see if any have been missed?

> **Vignettes**
> *"Estimating the proportion of cases that are ascertained during an outbreak":*
> - Another way to visualize the results to remove the effect of the `severity_baseline` argument is to plot the relative ratios between the countries so that the baseline severity cancels out, ruling out its effect.

I haven't understood this suggestion - how would you prefer the estimates to be treated? Do you mean I should scale all countries by the lowest value?

> *"Calculating a static, delay-adjusted estimate of disease severity":*
> - I'm not convinced that Fig 3 and 7 are necessary. It's not clear what it adds to the text. I would suggest to remove it. It looks like something that belongs in {epiparameter} rather than {cfr}.

Thanks - I've removed these figures and other plots showing the delay distributions. Could you also come in on issue #65. @CarmenTamayo initially raised the issue - would you like to weigh in as well?

>- I would suggest moving the section "Details: Adjusting for delays between two time series" to a theory vignette like in [bpmodels](https://epiverse-trace.github.io/bpmodels/articles/theoretical_background.html) and the details section of the respective functions. To be honest, it looks good here and seems to fit the context. Another way to go about it would be to hyperlink to the theory vignette at the beginning and end of this vignette like you have for the Getting Started vignette.

I think that it might be good to have the details within the function vignette for now, for quick reference. Happy to hear thoughts from @adamkucharski and others on whether we should take the single-theory-vignette approach? I've added these to the Details sections of function docs.

>- I would suggest replacing the generic use of "known outcomes" with the specifics. I think in this vignette, it is the known death outcomes.

I have replaced "known outcomes" with "known death outcomes" within vignettes where we calculate severity in terms of CFR. In the _Details_ sections however, I have retained "known outcomes" as the section refers to other severity measures as well.

>**"Estimating how disease severity varies over the course of an outbreak":**
>- Title seems long. Suggest shortening to, for example, "Estimating time-varying CFRs"

I'm not sure the shorter title would be clearer or as clear - let's keep this as it is for now and revisit it in case users find it confusing.

>- Same as before, I'm not convinced that Fig 3 is necessary. It's not clear what it adds to the text. I would suggest to remove it.

Removed.

>- Suggest adding a legend to Fig 4 and 5 indicate what the grey and tomato lines/ribbons represent.

Added.

>- Also, what is going on before April in Figs 4 and 5? The ribbons look weird and are hard to interpret.

My guess is that the data are pretty bad for this period and don't play well with the burn-in time or the rolling median estimate. I can remove dates prior to April 2020, but I imagine users will understand why the plot looks this way.

>- Refer to my earlier comment above on moving the section "Details: Adjusting for delays between two time series".

I think this section should stay here for now for quick reference, but happy to make a switch too; see above.

> Package version is missing the patch number, i.e., currently `0.1` instead of `0.1.0`. I would suggest to addding it to conform with the rest of our packages.

This has been added.

> Might be worthwhile to provide more descriptive contexts for the tests, for example, `test_that("multiplication works", ...`.

I have more recently moved to writing test descriptions as `"function_name: behaviour"`, as I find it easier to determine which function is failing. But this might be brittle as we're already considering function name changes, so I have replaced this format with a more descriptive one.

jamesmbaazam commented:
> This is a fair point. @jamesmbaazam - you've suggested making correcting for delays the default, but this suggestion would conflict with that - do you have thoughts here as well?

I think @Degoot-AM's point (RE: "What are the chances that the user will provide a value for the `epidist` parameter while ignoring to correct for the delays? ") is brilliant. 

It makes sense to remove the delay correction argument and make correcting delays contingent on value of epidist. I see you've made that change and I agree with it.

pratikunterwegs commented:
> It makes sense to remove the delay correction argument and make correcting delays contingent on value of epidist. I see you've made that change and I agree with it.

Thanks @jamesmbaazam - the change is still only proposed and hasn't been pushed. I'm also reducing the use of logical flags in other instances, such as `smooth_inputs` in `estimate_time_varying()` (proposed name change to `cfr_time_varying()`.

jamesmbaazam commented:
> Would values > 1.0 have some special interpretation, or simply be a mistake? 

Generally, comparison to a baseline (by division or some other operation) helps to make relative statements about the measured quantity. Maybe, a future enhancement would be to return qualitative expressions of the comparison, for example, `ifelse(baseline/measured quantity > 1, "CFR increasing", "CFR decreasing")`. 

The implemented warning should be a good signal to the user of what is happening, in case, the internal operation is undesirable.

jamesmbaazam commented:
> Thanks, I have added the details for these methods from the corresponding vignettes. Could you take a look and say whether this is sufficient?

It looks really good. I like it.

jamesmbaazam commented:
> Thanks for suggesting this - I'll bump this discussion to the top and ask for more feedback. I would prefer a naming convention of `cfr_x` where `x` is the method. This helps with communicating the functionality as well as with autocomplete.

The prefix also reads fine.

jamesmbaazam commented:
> I agree with this and I thought I had removed these, but I've removed all remaining instances as well now - could you please see if any have been missed?

Sure, I found the following:

- https://github.com/epiverse-trace/cfr/blob/66bfa793e6d2a6f51f8de5a1ad6056b74c54350a/vignettes/estimate_static_severity.Rmd#L80C40-L80C40
- https://github.com/epiverse-trace/cfr/blob/66bfa793e6d2a6f51f8de5a1ad6056b74c54350a/vignettes/estimate_ascertainment.Rmd#L113
- https://github.com/epiverse-trace/cfr/blob/66bfa793e6d2a6f51f8de5a1ad6056b74c54350a/vignettes/estimate_ascertainment.Rmd#L140

jamesmbaazam commented:
> I haven't understood this suggestion - how would you prefer the estimates to be treated? Do you mean I should scale all countries by the lowest value?

Yes, so I was suggesting that an alternative approach would be to choose one of the estimates as the new baseline, divide everything by that, and plot it. It would show the relative estimates, removing the impact of the baseline, i.e., $\dfrac{baseline}{estimate A}/\dfrac{baseline}{estimate B}=\dfrac{estimate B}{estimate A}$. I was just thinking out loud because I have a quibble with the baseline's effect. Its value determines how many estimates will be replaced by 1 in the end.

pratikunterwegs commented:
> A part of the reference is not rendering right, though. See last line of the attached. Same goes for estimate_time_varying.

This renders alright for me once the dev version is installed. I think this might just be an issue when it is temporarily loaded.

jamesmbaazam commented:
> Thanks - I've removed these figures and other plots showing the delay distributions. Could you also come in on issue #65. @CarmenTamayo initially raised the issue - would you like to weigh in as well?

I've commented on that issue [here](https://github.com/epiverse-trace/cfr/issues/65#issuecomment-1719292687). I think that the plots referred to here are a little more difficult to interpret and take away from the core message, compared to the plotted time series. 

I've seen a similar thing done in an [epinowcast vignette](https://package.epinowcast.org/dev/articles/single-timeseries-rt-estimation.html#specifying-the-model-using-epinowcastenw_expectation) but there, it was necessary to plot it because it was preceded by a detailed explanation of its setup.

jamesmbaazam commented:
> I think that it might be good to have the details within the function vignette for now, for quick reference. Happy to hear thoughts from @adamkucharski and others on whether we should take the single-theory-vignette approach? I've added these to the Details sections of function docs.

Agreed. I think they look nicer in the individual vignettes. I later left a comment to that effect.

jamesmbaazam commented:
> This is an interesting idea, and I think it would not take too much effort to implement. Still, I think it's good to have the conceptual separation between preparing data for CFR calculation, and the estimation step, reflected in the code for now. Happy to hear more thoughts on this from others.

The following linked issue in {serofoi} might be relevant to this conversation. @zmcucunuba also thinks that it makes sense to have two separate routines for preparing the data and modelling. See https://github.com/epiverse-trace/serofoi/issues/61.

pratikunterwegs commented:
Thanks again everyone for your feedback. The proposed and (some) scheduled changes have now been implemented. This PR will be retargeted to `main` and merged tomorrow morning. Please leave any final suggestions by then.

Bisaloo commented:
> These functions now have the epidist argument set to missing by default, the previous default was `NULL`

What's the motivation for this change? `NULL` probably signals more clearly that this argument is optional. 

This is discussed in the tidyverse design guide as well: https://design.tidyverse.org/required-no-defaults.html

pratikunterwegs commented:
> > These functions now have the epidist argument set to missing by default, the previous default was `NULL`
> 
> What's the motivation for this change? `NULL` probably signals more clearly that this argument is optional.
> 
> This is discussed in the tidyverse design guide as well: https://design.tidyverse.org/required-no-defaults.html

This has been changed to NULL by default in https://github.com/epiverse-trace/cfr/pull/78/commits/a5cdca3014261d78c9e54764edd40e6e2639e4de

pratikunterwegs commented on vignettes/estimate_time_varying_severity.Rmd:
````diff
@@ -0,0 +1,372 @@
+```
+
+Then, we plot case incidence data with following command.
+Further plotting commands are hidden for brevity.
````
I think we should show the first instance. Further instances are folded.

pratikunterwegs commented on R/estimate_static.R:
```diff
@@ -0,0 +1,141 @@
+#' @param correct_for_delays A boolean flag indicating whether the user wishes
+#' to correct for the delay between case detection and death. FALSE corresponds
+#' to a naive severity being calculated, TRUE corresponds to the user
+#' calculating a corrected severity
```
This documentation has been changed

pratikunterwegs commented on R/known_outcomes.R:
```diff
@@ -0,0 +1,101 @@
+#' parameterised with disease-specific parameters before it is supplied here.
+#' A typical example would be a symptom onset to death delay distribution.
+#'
+#' @return A data.frame with the columns in `data`, and with two additional
```
I've shifted to using `<data.frame>` throughout

pratikunterwegs commented on R/known_outcomes.R:
```diff
@@ -0,0 +1,101 @@
+  # time units are used) since the first case was reported
+  case_times <- as.numeric(data$date - min(data$date, na.rm = TRUE),
+    units = "days"
+  ) + 1
```
Fixed elsewhere

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+#' Alternatively, a sensible value might be 7, to disregard the first week of
+#' cases and deaths.
+#' To consider all case data including the start of the time-series, set this
+#' argument to 1.
```
Sure - another case is if you want to include all days. I've allowed this value to have a lower limit of 0 (previously 1) to include all days. You can now exclude N days by passing N, which does make a lot more sense than the previous setup.

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+#' procedure before calculating the time-varying severity.
+#' Useful for noisy time-series or time-series with strong reporting
+#' (e.g., weekend) effects.
+#' The default value is 1 for _no smoothing_. Values > 1 apply smoothing.
```
This documentation has been updated.

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+#' a case fatality ratio over time is estimated
+#'
+#' @inheritParams estimate_static
+#' @param burn_in_value The number of time-points (typically days) to disregard
```
I've renamed `burn_in_value` to `burn_in`, see https://github.com/epiverse-trace/cfr/pull/78/commits/9e755ffc9dd12a8dbc0b80a9adf28aecc207a483

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+#' @inheritParams estimate_static
+#' @param burn_in_value The number of time-points (typically days) to disregard
+#' at the start of the time-series, if a burn-in period is desired.
+#' The default value is set to the mean of the central spread of the `<epidist>`
```
Yes, this has been updated earlier

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+#'
+#' @description Calculates how the severity of a disease changes over time,
+#' corrected for a user-specified delay. If cases are supplied, and the delay
+#' distribution representing the delay between case detection and death, then
```
Updated in a different commit

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+#' data("covid_data")
+#' df_covid_uk <- covid_data[covid_data$country == "United Kingdom", ]
+#'
+#' df_covid_uk_subset <- subset(df_covid_uk, date <= "2020-05-31")
```
I'm removing the subsetting as it's only to speed up processing by a bit

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+                                  smooth_inputs = FALSE,
+                                  smoothing_window = 1,
+                                  correct_for_delays = FALSE) {
+  # TODO input checking
```
This has been removed

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+  checkmate::assert_logical(correct_for_delays, len = 1L, any.missing = FALSE)
+  checkmate::assert_integerish(burn_in_value, lower = 1, len = 1L)
+  checkmate::assert_data_frame(data)
+  checkmate::assert_logical(correct_for_delays, len = 1L)
```
This has also been removed

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+
+  stopifnot(
+    "Case data must contain columns `cases` and `deaths`" =
+      (all(c("cases", "deaths") %in% colnames(data))),
```
Removed - I'm relying on check_names() now

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+    df_temp$cases <- stats::runmed(data$cases,
+      k = smoothing_window,
+      endrule = "keep"
+    )
```
Updated in a different commit

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+    df_temp$deaths <- stats::runmed(data$deaths,
+      k = smoothing_window,
+      endrule = "keep"
+    )
```
Updated in a different commit

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+
+  case_times <- as.numeric(df_temp$date - min(df_temp$date, na.rm = TRUE),
+    units = "days"
+  ) + 1
```
Updated in a different commit

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+    pmf_vals <- stats::density(
+      epidist,
+      at = seq(from = 0, to = nrow(data) - 1L)
+    )
```
I don't understand where this is duplicated - the method is indeed defined in {epiparameter} already

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+          rev(delay_pmf_eval)
+
+        # return rounded sum of known_onsets_current
+        round(sum(known_onsets_current, na.rm = TRUE))
```
I'm not entirely certain about the logic behind this rounding, but I expect this was introduced by Tim R in the initial implementation. My best guess is that it yields integerish numbers of onsets. If you can see a specific issue with placing it here, I can change the implementation

pratikunterwegs commented on R/estimate_time_varying.R:
```diff
@@ -0,0 +1,188 @@
+  for (i in indices) {
+    # handle case where deaths are fewer than non-zero known outcomes
+    if (df_temp$deaths[i] <= df_temp$known_outcomes[i] &&
+      isTRUE(df_temp$known_outcomes[i] > 0)) {
```
See above - the issue is how to handle fractional known outcomes when correcting for delays

pratikunterwegs commented on R/estimate_severity.R:
```diff
@@ -0,0 +1,63 @@
+                              total_outcomes,
+                              poisson_threshold = 100) {
+  # Add input checking for single numbers
+  checkmate::assert_count(total_cases)
```
`assert_count` already expects a non negative value

pratikunterwegs commented on R/estimate_severity.R:
```diff
@@ -0,0 +1,63 @@
+#' @param total_deaths The total number of deaths observed over the period of an
+#' outbreak of interest. The total number of deaths must be less than or equal
+#' to the total number of cases.
+#' @param total_outcomes The total number of outcomes estimated observed over
```
Corrected to "expected to be observed"

pratikunterwegs commented on R/estimate_severity.R:
```diff
@@ -0,0 +1,63 @@
+#' Estimate the corrected case fatality rate
+#'
+#' @description Estimate the maximum likelihood estimate and 95% confidence
```
Changed overall

pratikunterwegs commented on R/estimate_ascertainment.R:
```diff
@@ -0,0 +1,147 @@
+                                   ),
+                                   smooth_inputs = FALSE,
+                                   smoothing_window = 1,
+                                   correct_for_delays = FALSE,
```
This has been fixed by making delay correction conditional on epidist

pratikunterwegs commented on R/estimate_rolling.R:
```diff
@@ -0,0 +1,127 @@
+#'     epidist = onset_to_death_ebola
+#'   )
+#' )
+#'
```
This has been separated

pratikunterwegs commented on R/prepare_data.R:
```diff
@@ -0,0 +1,166 @@
+#'
+#' Keeping `NA`s will cause downstream issues when calling functions such as
+#' [estimate_static()] on the data, as they cannot handle `NA`s.
+#' Setting `fill_NA = TRUE` resolves this issue, but must be a conscious choice.
```
I will make this a separate issue and let people come in on it - I do think that users should be made to consciously choose.

pratikunterwegs commented on R/estimate_rolling.R:
```diff
@@ -0,0 +1,127 @@
+  } else {
+    # prepare cumulative sums
+    cumulative_cases <- cumsum(data$cases)
+    cumulative_deaths <- cumsum(data$deaths)
```
These have been moved outside the ifelse

pratikunterwegs commented on man/.DS_Store:
This has been removed

pratikunterwegs commented on vignettes/estimate_ascertainment.Rmd:
```diff
@@ -0,0 +1,294 @@
+::: {.alert .alert-primary}
+## Use case {-}
+
+We wish to estimate one of the typical severity quantities used in epidemiology -- the *case fatality risk (CFR)*, *infection hospitality risk (IFR)* or  *hospitalisation fatality risk (HFR)* -- using a method that uses a time series of cases, infections or hospitalisations (respectively) and deaths, _while correcting for the delay in reporting the outcomes of cases_.
```
I've renamed these to risks per Adam's advice

pratikunterwegs commented on vignettes/estimate_ascertainment.Rmd:
```diff
@@ -0,0 +1,294 @@
+library(purrr)
+library(scales)
+library(forcats)
+library(ggplot2)
```
I would prefer to load the individual packages instead actually

pratikunterwegs commented on vignettes/estimate_ascertainment.Rmd:
```diff
@@ -0,0 +1,294 @@
+We exclude four countries which only provide weekly data (with zeros for dates in between) --- France, Germany, Spain and Ukraine.
+
+We adopt data science tools to iteratively apply the `estimate_ascertainment()` function across data grouped by country.
+We refer the user to the book [R for Data Science](https://r4ds.had.co.nz/) for a better explanation of some of the code used here, including from the packages in [the Tidyverse](https://www.tidyverse.org/).
```
I don't think so - there's only so far we can go in educating users on data wrangling within our package vignettes. We are pointing them to a very good source of help.

pratikunterwegs commented on vignettes/estimate_ascertainment.Rmd:
```diff
@@ -0,0 +1,294 @@
+
+This known disease severity is passed to the `severity_baseline` argument in `estimate_ascertainment()`, and forms the denominator in the resulting under-ascertainment calculation.
+
+We assume that the 'true' CFR of Covid-19 is 0.014.
```
Good question - I'm not sure where this number comes from but my best bet is this paper: https://www.thelancet.com/journals/laninf/article/PIIS1473-3099(20)30243-7/fulltext
